src/test/java/com/uid2/admin/managers/KeysetManagerTest.java

```diff
@@ -0,0 +1,133 @@
+import static org.mockito.Mockito.*;
+
+public class KeysetManagerTest {
+    protected AutoCloseable mocks;
+
+    @Mock protected KeysetStoreWriter keysetStoreWriter;
+    @Mock protected RotatingKeysetProvider keysetProvider;
+    @Mock
+    protected IKeysetKeyManager keysetKeyManager;
+    @Mock protected KeysetSnapshot keysetSnapshot;
+
+    @BeforeEach
+    public void setupMocks() {
+        mocks = MockitoAnnotations.openMocks(this);
+        when(keysetProvider.getSnapshot()).thenReturn(keysetSnapshot);
+    }
+
+    @AfterEach
+    public void teardown() throws Exception {
+        mocks.close();
+    }
+
+    protected void setKeysets(Map<Integer, Keyset> keysets) {
+        when(keysetSnapshot.getAllKeysets()).thenReturn(keysets);
+    }
+
+    @Test
+    public void testAddOrReplaceKeyset() throws Exception{
+        KeysetManager keysetManager = new KeysetManager(keysetProvider, keysetStoreWriter,
+                keysetKeyManager, true);
+
+        Map<Integer, Keyset> keysets = new HashMap<Integer, Keyset>() {{
+                put(1, KeysetUtil.createDefaultKeyset(3, 1));
+                put(2, KeysetUtil.createDefaultKeyset(4, 2));
+                put(3, KeysetUtil.createDefaultKeyset(5, 3));
+                put(4, KeysetUtil.createDefaultKeyset(6, 4));
+            }};
+
+        setKeysets(keysets);
+
+        // add new keyset
+        Keyset keyset1 = KeysetUtil.createDefaultKeyset(7, 5);
+        keysetManager.addOrReplaceKeyset(keyset1);
+        assertTrue(keysets.containsKey(5));
+        assertTrue(keyset1.equals(keysets.get(5)));
+        verify(keysetStoreWriter).upload(mapOfSize(5), isNull());
+
+        //Reset between tests so verify works
+        reset(keysetStoreWriter);
+
+        // add existing Keyset
+        Keyset keyset2 = new Keyset(5, 7, "newKeyset", Set.of(1, 2, 3), 1 , true, true);
+        keysetManager.addOrReplaceKeyset(keyset2);
+        assertTrue(keysets.containsKey(5));
+        assertFalse(keyset1.equals(keysets.get(5)));
+        assertTrue(keyset2.equals(keysets.get(5)));
+        verify(keysetStoreWriter).upload(mapOfSize(5), isNull());
+    }
+
+    @Test
+    public void testCreateKeysetForClient() throws Exception {
+        KeysetManager keysetManager = new KeysetManager(keysetProvider, keysetStoreWriter,
+                keysetKeyManager, true);
+
+        Map<Integer, Keyset> keysets = new HashMap<Integer, Keyset>() {{
+            put(1, KeysetUtil.createDefaultKeyset(3, 1));
+            put(2, KeysetUtil.createDefaultKeyset(4, 2));
+            put(3, KeysetUtil.createDefaultKeyset(5, 3));
+            put(4, KeysetUtil.createDefaultKeyset(6, 4));
+        }};
+
+        setKeysets(keysets);
+        // Sharer makes an empty list
+        ClientKey sharer = new ClientKey("", "", "",  "", Instant.now(), Set.of(Role.SHARER), 7, false);
+        keysetManager.createKeysetForClient(sharer);
+        Keyset sharerKeyset = keysets.get(5);
```

Comment:
That "5" seems fragile and hard to read; we're relying on a particular keyset id assignment algorithm which might change. 
Can we rely on the return value from createKeysetForClient to get the keyset id?

Reply:
createKeysetForClient doesn't have a return value?

Reply:
maybe it doesn't now - but how can we be certain the new keyset id will be "5"?

Reply:
I used the same method that picks the next keysetID to set this now